### PR TITLE
Tools for LAMMPS documentation for developers with the "doxygen" documentation platform

### PR DESCRIPTION
**Pull request by numericalfreedom**
Tools for creating LAMMPS developer documentation with the "doxygen" documentation platform are provided.

        New directory: tools/doxygen

        New file:      tools/doxygen/Developer.dox.lammps
        New file:      tools/doxygen/Doxyfile.lammps
        New file:      tools/doxygen/doxygen.sh
        New file:      tools/doxygen/README

The Developer.dox.lammps file contains a slightly revised version of the
Developer.pdf file adopted to the LAMMPS "doxygen" documentation.

The Doxyfile.lammps file is a first proposal for a LAMMPS "doxygen"
documentation flavor and can be adjusted to specific requirements.

The "doxygen.sh" shell script generates the LAMMPS "doxygen"
documentation.

Detailed instructions can be found in the README file.

## Purpose

_Tools for LAMMPS documentation with the "doxygen" documentation platform are provided._

## Author(s)

_Nandor Tamaskovics, numericalfreedom AT googlemail.com_

## Backward Compatibility

_ No limitation _

## Implementation Notes

_ No influence on computation results. Tools should be applied to a freshly cloned LAMMPS source copy and avoided to launched in a production environment _

## Post Submission Checklist

_Please check the fields below as they are completed_
- [X] The feature or features in this pull request is complete
- [X] Suitable new documentation files and/or updates to the existing docs are included (README)
- [-] One or more example input decks are included
- [-] The source code follows the LAMMPS formatting guidelines

## Further Information, Files, and Links

_ Check the tools/doxygen directory and the README file in it. _

### CODE REVIEW

File: tools/doxygen/doxygen.sh [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  error=1 ;
+  echo "Cannot configure LAMMPS sources - Please run doxygen.sh from the LAMMPS root directory."
+ fi ;
+
+
+if [ ${error} == 0 ] ;
+  then
+    if [[ ! -x ${FIG2DEV} ]] ;
+      then
+        error=1 ;
```
Comment (junghans) [inline tools/doxygen/doxygen.sh]: I would do an `exit 1` here and remove all the `[ ${error} = 0 ]` below.

File: tools/doxygen/doxygen.sh [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        echo "${FIG2DEV} not found - Please install ${FIG2DEV} for Your operating system." ;
+      fi ;
+  fi ;
+
+
+if [ ${error} == 0 ] ;
```
Comment (junghans) [inline tools/doxygen/doxygen.sh]: Use `-eq` instead of `==`.

File: tools/doxygen/doxygen.sh [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+
+if [[ -d ${LAMMPSSRC} && -d ${LAMMPSDEVELOPERDOC} && -d ${LAMMPSDOC} && -f ${LAMMPSDOCMAKE} && -d ${LAMMPSDOXYGEN} && -f ${LAMMPSDOXYFILE} ]] ;
+ then
+   cd ${LAMMPSSRC} ;
+   make no-all ;
+   cd ${LAMMPSBASE} ;
+ else
+  error=1 ;
+  echo "Cannot configure LAMMPS sources - Please run doxygen.sh from the LAMMPS root directory."
```
Comment (junghans) [inline tools/doxygen/doxygen.sh]: Error message should go to stderr, add a `>&2` at the end of the line.

File: tools/doxygen/doxygen.sh [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+      fi ;
+  fi ;
+
+
+if [ ${error} == 0 ] ;
+  then
+    if [[ -d ${LAMMPSDOC} && -f ${LAMMPSDOCMAKE} ]] ;
+      then
+        cd ${LAMMPSDOC} ;
+        if [ -d ${LAMMPSDOCTMP} ] ;
```
Comment (junghans) [inline tools/doxygen/doxygen.sh]: You are check for a directory ending in `*`here. Is that correct?

File: tools/doxygen/doxygen.sh [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+      echo "Cannot build LAMMPS native documentation - Please run doxygen.sh from the LAMMPS root directory." ;
+    fi ;
+  fi ;
+
+
+if [ ${error} == 0 ] ;
+  then
+    if [[ ${error} == 0 && -d ${LAMMPSDOCHTML} ]] ;
+      then
+        if [ -d ${LAMMPSDOXYGENDOCMANUAL} ] ;
```
Comment (junghans) [inline tools/doxygen/doxygen.sh]: Use `[[...]]` or quote the variable.

File: tools/doxygen/doxygen.sh [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+      fi ;
+  fi ;
+
+
+if [ ${error} == 0 ] ;
+  then
+    if [[ -d ${LAMMPSSRC} && -f ${LAMMPSDOXYFILE} && -f ${LAMMPSDEVELOPERDOXFILE} ]] ;
+      then
+        ICS=' ';
+        read -r -a array < ${LAMMPSSRC}/version.h ;
```
Comment (junghans) [inline tools/doxygen/doxygen.sh]: That seems a bit complicated, something like `sed -e 's/.*"//' -e 's/".*//' -e 's/ //' version.h` should to it all at once.

File: tools/doxygen/doxygen.sh [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+LAMMPSDOXYGENDOCMANUAL="${LAMMPSDOXYGEN}/doc/html/Manual" ;
+
+LAMMPSDOXYFILE=${LAMMPSDOXYGEN}/Doxyfile.lammps ;
+DOXYFILE=${LAMMPSDOXYGEN}/Doxyfile ;
+
+LAMMPSDEVELOPERDOXFILE=${LAMMPSDOXYGEN}/Developer.dox.lammps ;
+DEVELOPERDOXFILE=${LAMMPSDOXYGEN}/Developer.dox ;
+
+FIG2DEV=`which fig2dev` ;
+DOXYGEN=`which doxygen` ;
```
Comment (junghans) [inline tools/doxygen/doxygen.sh]: In bash you can use `type -p` instead of `which`.

File: tools/doxygen/doxygen.sh [hunk 8]
```diff
@@ -0,0 +1,127 @@
+#!/bin/bash
+
+LAMMPSBASE=${PWD} ;
+LAMMPSSRC="${LAMMPSBASE}/src" ;
+LAMMPSDOC="${LAMMPSBASE}/doc" ;
+LAMMPSDOCHTML="${LAMMPSDOC}/html" ;
+LAMMPSDEVELOPERDOC="${LAMMPSBASE}/doc/src/Developer" ;
+LAMMPSDOCTMP="/tmp/lammps-docs-*" ;
```
Comment (junghans) [inline tools/doxygen/doxygen.sh]: See the comment below, but the `*` will never get expanded.

File: tools/doxygen/doxygen.sh [hunk 9]
```diff
@@ -0,0 +1,127 @@
+#!/bin/bash
+
+LAMMPSBASE=${PWD} ;
```
Comment (junghans) [inline tools/doxygen/doxygen.sh]: All the `;` are not needed.

Comment (numericalfreedom) [reply tools/doxygen/doxygen.sh]: Not needed but syntactically correct. Makes code argumentation more clear. Bash likes them.

File: tools/doxygen/doxygen.sh (hunk 7 above)
Comment (numericalfreedom) [reply tools/doxygen/doxygen.sh]: `which ...` works also fine.

File: tools/doxygen/doxygen.sh (hunk 2 above)
Comment (numericalfreedom) [reply tools/doxygen/doxygen.sh]: == is better readable, works fine.

File: tools/doxygen/doxygen.sh (hunk 1 above)
Comment (numericalfreedom) [reply tools/doxygen/doxygen.sh]: I simplify the shell script with exits.

File: tools/doxygen/doxygen.sh (hunk 6 above)
Comment (numericalfreedom) [reply tools/doxygen/doxygen.sh]: I am not too much a "sed" expert.

File: tools/doxygen/doxygen.sh (hunk 5 above)
Comment (numericalfreedom) [reply tools/doxygen/doxygen.sh]: I have removed double quotes everywhere, where they would be obsolete.

**Comment by numericalfreedom** [comment]
Precisions to the tools/doxygen/doxygen.sh have been added and shell script has been tested.

**Pull request merged**